src/ReverseProxy/Service/Management/ItemManagerBase.cs

```diff
@@ -98,9 +96,9 @@ protected virtual void OnItemChanged(T item, bool added)
         protected virtual void OnItemRemoved(T item)
         {}
 
-        private void UpdateSignal()
+        private void UpdateSnapshot()
         {
-            _signal.Value = _items.Select(kvp => kvp.Value).ToList().AsReadOnly();
+            _snapshot = _items.Select(kvp => kvp.Value).ToList().AsReadOnly();
```

Comment:
Why are we still doing this?

Reply:
Because _items isn't thread safe so we create a snapshot under the lock. What did you have in mind? Switch to ConcurrentDictionary?

Reply:
I think making _items a ConcurrentDictionary makes some sense. Even if we do that, we'd want to take snapshots assuming that we read the values more than we update them. Either way, couldn't this be rewritten to the following?

```suggestion
            _snapshot = GetItems().AsReadOnly();
```

This makes me wonder why GetItems() exists.

Reply:
We don't want to allocate a new list when items change, so we need something concurrent or we need to lock and mutate.

Reply:
I actually expect more changes than reads here. E.g. Reads are only triggered when we're pretty sure something has already changed, and we may change more than one thing before the read happens.

Note we end up making a copy regardless when the health checks are enabled because we filter the list. I'll try changing to a ConcurrentDictionary since the contention should be relatively low (e.g. not updated per-request, and only on background threads).

Reply:
That's bad, but we can do these changes in stages. We don't want GC pressure when clusters frequently update (we don't have any performance scenarios that test those sorts of scenarios).

Reply:
I *just* re-read the code and remembered how these updates happen. It seems like we'd be better off with an immutable dictionary instead of the whole IItemManager thing. We can change the internal implementation for now then look to change calling code once we're done with that.

Reply:
#513 to track cleaning up the IItemManager types.